brainiak/utils/fmrisim.py

```diff
@@ -1137,86 +1160,128 @@ def _calc_snr(volume,
 
     Parameters
     ----------
 
-    volume : 4d array, float
+    volume : 4d array or 1d array, float
         Take a volume time series to extract the middle slice from the
-        middle TR
+        middle TR. Can also accept a one dimensional time course (mask input
+        is then ignored).
 
     mask : 3d array, binary
         A binary mask the same size as the volume
 
     auto_reg_order : int
-        What order of the autoregression do you want to pull out
+        What order of the autoregression do you want to estimate
 
+    sample_num : int
+        How many voxels would you like to sample to calculate the AR values.
+        The AR distribution of real data is approximately exponential maxing
+        at 1. From analyses across a number of participants, to get less
+        than 1% standard deviation of error from the true mean it is
+        necessary to sample at least 1000 voxels.
 
     Returns
     -------
 
+    auto_reg_rho : list of floats
+        Rho of a specific order for the autoregression noise in the data
 
-    sfnr : float 
-        The SFNR of the volume (mean brain activity divided by  temporal
-        variability in the averaged non brain voxels)  
+    na_rho : list of floats
+        Moving average of a specific order for the data
 
-    auto_reg_sigma : float
-        A sigma of the autoregression in the data
+    """
 
-    drift_sigma : float
-        Sigma of the drift in the data
+    # Pull out the non masked voxels
+    if len(volume.shape) > 1:
+        brain_timecourse = volume[mask > 0]
+    else:
+        # If a 1 dimensional input is supplied then reshape it to make the
+        # timecourse
+        brain_timecourse = volume.reshape(1, len(volume))
 
-    """
+    # Identify some brain voxels to assess
+    voxel_idxs = list(range(brain_timecourse.shape[0]))
+    np.random.shuffle(voxel_idxs)
+
+    # If there are more samples than voxels, take all of the voxels
+    if len(voxel_idxs) < sample_num:
+        sample_num = len(voxel_idxs)
 
-    # Calculate sfnr and convert from memmap
-    sfnr = _calc_sfnr(volume,
-                      mask,
-                      )
+    auto_reg_rho_all = np.zeros((sample_num, auto_reg_order))
+    ma_all = np.zeros((sample_num, ma_order))
+    for voxel_counter in range(sample_num):
 
-    # Calculate the time course of voxels within the brain
-    timecourse = np.mean(volume[mask > 0], 0)
-    demeaned_timecourse = timecourse-timecourse.mean()
+        # Get the timecourse and demean it
+        timecourse = brain_timecourse[voxel_idxs[voxel_counter], :]
+        demeaned_timecourse = timecourse - timecourse.mean()
 
-    # Pull out the AR values (depends on order)
-    auto_reg_sigma = ar.AR_est_YW(demeaned_timecourse, auto_reg_order)
-    auto_reg_sigma = np.sqrt(auto_reg_sigma[1])
+        # Pull out the ARMA values (depends on order)
+        try:
+            model = ARMA(demeaned_timecourse, [auto_reg_order, ma_order])
```

Comment:
This would be great if I could avoid this because this is the most time consuming part of the whole code. However, this line seems necessary since the model also takes in the data (demeaned_timecourse) which changes on each iteration right